### ISSUE TITLE
Using spaCy vectors in PyTorch: Vectors.add() and OOV

### ISSUE THREAD

**Issue by BramVanroy** [explosion/spaCy#3335]
I am trying to add the vectors of `en_vectors_weg_lg` as an [embedding laye](https://pytorch.org/docs/stable/nn.html#embedding)r to PyTorch, which can be done with the [from_pretrained](https://pytorch.org/docs/stable/nn.html#torch.nn.Embedding.from_pretrained) method. The expected input to build this layer is the embedding weights. I though I could use the following presentation for that:

```python
import spacy
nlp = spacy.load('en_vectors_web_lg')
embed_weights = torch.FloatTensor(nlp.vocab.vectors.data)
embed_layer = nn.Embedding.from_pretrained(embed_weights, freeze=True)
````

Then I can use the key2row dict to convert my input tokens into IDs that the embedding layer can use.

```python
import spacy

nlp = spacy.load('en_vectors_web_lg')

embed_weights = torch.FloatTensor(nlp.vocab.vectors.data)
embed_layer = nn.Embedding.from_pretrained(embed_weights, freeze=True)

cat_id = nlp.vocab.strings[u'cat']
print(f"ID: {cat_id}")
cat_row = nlp.vocab.vectors.key2row[cat_id]
print(f"row: {cat_row}")

# cat_embedding is a Tensor representation of cat_vector
cat_vector = nlp.vocab.vectors[cat_id]
print(f"vector: {cat_vector}")
cat_embedding = embed_layer(torch.LongTensor([cat_row]))
print(f"embedding: {cat_embedding}")
```

All of this works perfectly. The question that I am having is related to dealing with OOV, which is an issue in itself. I found that you can use a random string in `nlp.vocab.strings` and will always (?) return a value. (The value for `nlp.vocab.strings[u'azefdgdfg46dfg64dfrdf']` is 15688863475460103551). I am not sure why. But this number cannot be found in the key2row dict, which is no surprise.

My basic idea was to add a new token with a zero-vector representation, e.g. `@UNK@`, and when there is a KeyError, use the the key2row ID of this new token. I tried `add()` like so

```python
nlp.vocab.vectors.add('@pad@', vector=np.zeros(nlp.vocab.vectors.shape[1]))
```

But I got the following error

> [E060] Cannot add new key to vectors: the table is full. Current shape: (1070971, 300).

It would seem odd to me that you need to do some resizing before you can add a new value so I feel I am missing a point. The documentation is quite brief with this and would benefit from an expansion. I would like to do this, but I don't quite understand what the correct approach is to this. Should I first resize, then add? Should I first add a new row, then add to that row? Code examples welcome.

_As a side note, I would also add a `@PAD@` token for padded sentences, but it seems wrong to have the same vector representation for OOV and padding tokens. Any tips welcome._

## Which page or section is this issue related to?
[Vectors.add](https://spacy.io/api/vectors#add)

**Comment by honnibal** [comment]
If you want a method that can automatically resize the vectors table, you should try `nlp.vocab.vectors.add('@pad@', vector=np.zeros(nlp.vocab.vectors.shape[1]))`

Here's a few details about why these things work the way they do.

The `Vectors` class wants to make sure the underlying vectors data is stored in a contiguous table, so we need to maintain a mapping of keys to rows in this table. We want to avoid having to resize the vectors data on every insertion, as that would be really slow. So, we keep track of which rows haven't been assigned to any keys. Multiple keys may be mapped to the same row, so that if we know some words are synonyms, we can represent that cheaply. For instance, you could have some entries case insensitive this way, by just adding another key.

The `nlp.vocab` class provides a higher-level interface to the vectors. It manages how vectors are associated to words. This lets you write `set_vector()`, and it takes care of resizing if necessary.

**Comment by honnibal** [comment]
By the way, if you want to use the vectors in PyTorch, you might want to have an efficient way to get the rows of the tokens of `Doc` object. The fastest way to do this is a bit quirky currently. After setting your vectors, call `spacy.vocab.link_vectors_to_models(nlp)`. You'll then be able to do `doc.to_array([ID])` to get a numpy array with the vector rows for each token in your document.

**Comment by BramVanroy** [comment]
Is it possible that you made a copy-and-paste error? The line of code that you provided is the same as the one that I gave (and it results in an error). 

I don't completely understand how to quickly get a Doc's tokens' vector representation, though. I tried the following:

```python
import spacy
from spacy.attrs import ID

nlp = spacy.load('en_core_web_lg')
w2v = spacy.load('en_vectors_web_lg')

spacy.vocab.link_vectors_to_models(w2v.vocab)

doc = nlp(u'hello world')
print(doc.to_array([ID]))
# [6066  272]
```

But I don't understand what this does. ID is the row ID of the token in the vector table (and is thus not present when a model does not have vectors), correct? So we can load a new vector model, and the ID will then be the one from the vector model? I don't quite understand how `spacy.vocab.link_vectors_to_models(w2v.vocab)` knows that I want to change the vectors of `nlp`. How does spacy know which spaCy instance the new vectors should be linked to?

Thanks for the help so far!

**Comment by honnibal** [comment]
Sorry, yes. Fixed my example.

As for the `link_vectors_to_models(nlp.vocab)`: The `Lexeme` objects have an attribute that the row ID is stored in. Confusingly this attribute is accessed as `lex.rank` in the Python API. The attribute name `ID` is unideal.

**Comment by BramVanroy** [comment]
But Lexeme object are unique per model, right? So how come that we don't have to pass the nlp model to `link_vectors_to_models()`? In my example above, I load the vectors from `en_vectors_weg_lg` (1) and want to use them inside the model `en_core_web_lg` (2). So when I use `link_vectors_to_models()` how does spaCy know I want to load 1 into **2**, since it is not referenced in that method call.

**Comment by honnibal** [comment]
Well the vectors has this mapping `key2row`, right? The `lexeme.orth` is the key there. All we're doing is storing the `row` on the lexeme, so that we can retrieve it easily without looking up the dict on every token.
